shared/util_linux.go

```diff
@@ -193,6 +194,73 @@ int shiftowner(char *basepath, char *path, int uid, int gid) {
 	return 0;
 }
 
+int shiftacl(char *basepath, char *path, int uid, int gid, int new_uid, int new_gid) {
+	struct stat sb;
+	acl_t acl;
+	int entry_id;
+
+	acl = acl_get_file(path, ACL_TYPE_ACCESS);
+	if (acl == NULL) {
+		return 0;
+	}
+
+	if (lstat(path, &sb) < 0) {
+		perror("stat failed");
+		acl_free(acl);
+		return -1;
+	}
+	if (S_ISLNK(sb.st_mode)) {
+		acl_free(acl);
+		return 0;
+	}
+
+
+	for (entry_id = ACL_FIRST_ENTRY; ; entry_id = ACL_NEXT_ENTRY) {
+		acl_entry_t ent;
+		acl_tag_t tag;
+		id_t *id_p;
+		id_t new_id;
+		int update_acl = 0;
+
+		if (acl_get_entry(acl, entry_id, &ent) != 1) {
+			break;
+		}
+
+		if (acl_get_tag_type(ent, &tag) == -1) {
+			perror("failed acl_get_tag_type");
+			acl_free(acl);
+			return 1;
+		}
+
+		id_p = acl_get_qualifier(ent);
+		if (id_p == NULL) {
+			continue;
+		}
+		switch(tag) {
+			case ACL_USER:
+				new_id = *id_p - uid + new_uid;
+				update_acl = 1;
+				break;
+
+			case ACL_GROUP:
+				new_id = *id_p - gid + new_gid;
+				update_acl = 1;
+				break;
+		}
+
+		if (update_acl) {
+			acl_set_qualifier(ent, &new_id);
+			if (acl_set_file(path, ACL_TYPE_ACCESS, acl) == -1) {
+				perror("acl_set_file failed");
+				acl_free(id_p);
+				return 1;
+			}
+		}
+		acl_free(id_p);
```

Comment:
You'll likely need to `acl_free(acl)` here too unless `acl_set_file()` claims the memory and frees ith with `acl_free(id_p)`.